modules/littlefs/zephyr_lfs_crc.c

```diff
@@ -0,0 +1,31 @@
+/*
+ * lfs util functions
+ *
+ * Copyright (c) 2017, Arm Limited. All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include "lfs_util.h"
+
+/* Use the LFS naive CRC implementation until it has been decided which CRC to
+ * use for LittleFS.
+ */
+
+/* Software CRC implementation with small lookup table */
+uint32_t lfs_crc(uint32_t crc, const void *buffer, size_t size)
+{
+	static const uint32_t rtable[16] = {
+		0x00000000, 0x1db71064, 0x3b6e20c8, 0x26d930ac,
+		0x76dc4190, 0x6b6b51f4, 0x4db26158, 0x5005713c,
+		0xedb88320, 0xf00f9344, 0xd6d6a3e8, 0xcb61b38c,
+		0x9b64c2b0, 0x86d3d2d4, 0xa00ae278, 0xbdbdf21c,
+	};
+
+	const uint8_t *data = buffer;
+
+	for (size_t i = 0; i < size; i++) {
+		crc = (crc >> 4) ^ rtable[(crc ^ (data[i] >> 0)) & 0xf];
+		crc = (crc >> 4) ^ rtable[(crc ^ (data[i] >> 4)) & 0xf];
+	}
+
+	return crc;
+}
```

Comment:
I don't understand the need to copy this file.

Why don't we just add https://github.com/littlefs-project/littlefs/blob/master/lfs_util.c to the build ?

Reply:
Oh, I see we are:

https://github.com/zephyrproject-rtos/zephyr/pull/41729/files#diff-01ac9904de8007b215faf5b86a7ece70d162e5a70244a85a26b2166bb4d8e95bR14

so what's the problem? The define ifdef's out the function?

Then maybe we can undefine it, or have a source file that #include's lfs_util.c.

It's unfortunate to copy crypto code like this due to it being difficult to track security bugs when the code is copied.

Reply:
Maybe littlefs will listen to reason so we don't have to deal with these hacks.

https://github.com/littlefs-project/littlefs/pull/647

Reply:
@SebastianBoe, I think the idea is to replace this with a call to zephyr's builtin crc32. It is just not done at the moment.

Reply:
> @SebastianBoe, I think the idea is to replace this with a call to zephyr's builtin crc32. It is just not done at the moment.

Exactly, as the commit message states:
https://github.com/zephyrproject-rtos/zephyr/pull/41729/commits/399c1b88e2c794a62b7602fa76edfd0635dddb26
> This allows to ensure identical behavior when moving glue code into
Zephyr repo, and allow us to change to an existing crc implementation
provided natively by Zephyr at a later time.

The crc implementation provided with LFS is very naive, as it even says directly in the code.
https://github.com/zephyrproject-rtos/zephyr/blob/399c1b88e2c794a62b7602fa76edfd0635dddb26/modules/littlefs/zephyr_lfs_crc.c#L9-L10

In my opinion this CRC **is not** suitable for production code.

Therefore we should switch to using a Zephyr CRC implementation as this issue describes:
https://github.com/zephyrproject-rtos/zephyr/issues/43369

However such work is outside the scope of this PR, as this PR intends to move glue code out of LittleFS and thus reduce LittleFS patching and make it easier to upmerge.

Reply:
also note, with this approach, when #43369 is completed, then this commit can easily be reverted and thus removing this file again which is the end goal.
https://github.com/zephyrproject-rtos/zephyr/pull/41729/commits/399c1b88e2c794a62b7602fa76edfd0635dddb26